letsencrypt/client/client.py

```diff
@@ -389,18 +345,57 @@ def init_csr(privkey, names, cert_dir):
 
     return le_util.CSR(csr_filename, csr_der, "der")
 
+
 # This should be controlled by commandline parameters
-def determine_authenticator(config):
+def determine_authenticator(all_auths):
     """Returns a valid IAuthenticator.
 
-    :param config: Configuration.
-    :type config: :class:`letsencrypt.client.interfaces.IConfig`
+    :param list all_auths: Where each is a tuple of the form
+        ('description', 'IAuthenticator', *options..) where IAuthenticator is a
+        :class:`letsencrypt.client.interfaces.IAuthenticator` object or class
+        and options are the parameters used to initialize the authenticator.
+
+    :returns: Valid Authenticator object or None
 
     """
-    try:
-        return configurator.ApacheConfigurator(config)
-    except errors.LetsEncryptNoInstallationError:
-        logging.info("Unable to determine a way to authenticate the server")
+    # Available Authenticator objects
+    avail_auths = []
+    # Error messages for misconfigured authenticators
+    errs = {}
+
+    for pot_auth in all_auths:
+        try:
+            # I do not think this a great solution but haven't come up with
+            # anything better yet... other than constricting init functions for
+            # authenticators
```

Comment:
Yup, this is definitely over-engineered... Why don't just pass instances of authenticators to `determine_authenticator`? "Explicit is better than implicit" ([PEP 20](https://www.python.org/dev/peps/pep-0020/))...

Also, let's just make all these `IAuthenticator` constructors accept `IConfig` as the first (and only?) parameter for simplicity. By the way, I was thinking about [plugin system based on setuptools entry_points](https://pythonhosted.org/setuptools/setuptools.html#dynamic-discovery-of-services-and-plugins), so that everyone could create their own authenticator and the client would automatically discover what's installed on the system. That's a bit off topic here, so you can let me know what your thoughts are over IRC.